src/jit/codegenxarch.cpp

```diff
@@ -7397,30 +7397,106 @@ void CodeGen::genPutArgStk(GenTreePtr treeNode)
     }
     else if (data->OperGet() == GT_FIELD_LIST)
     {
-        GenTreeFieldList* fieldList = data->AsFieldList();
-        // TODO-X86-CQ: If there are no gaps or padding in the layout, we can use a series of pushes.
-        // For now, always make space for the struct, and store each field at its offset.
+        GenTreePutArgStk* const putArgStk = treeNode->AsPutArgStk();
 
-        unsigned structSize = treeNode->AsPutArgStk()->getArgSize();
-        m_pushStkArg        = false;
-        inst_RV_IV(INS_sub, REG_SPBASE, structSize, EA_PTRSIZE);
-        genStackLevel += structSize;
-        for (; fieldList != nullptr; fieldList = fieldList->Rest())
+        GenTreeFieldList* const fieldList = data->AsFieldList();
+        assert(fieldList != nullptr);
+
+        m_pushStkArg      = false;
+        const int argSize = putArgStk->getArgSize();
+        assert((argSize % TARGET_POINTER_SIZE) == 0);
+
+        // Set the current field offset to the size of the struct arg (which must be a multiple of the target pointer
+        // size).
+        int currentOffset = argSize;
+
+        // If there are no GC references in this field list and the argument is either larger than 16 bytes or the
+        // argument contains padding, we will pre-adjust the stack pointer and use stores instead of pushes.
+        const bool preAdjustStack =
+            (putArgStk->gtNumberReferenceSlots == 0) &&
+            ((argSize > 16) || (putArgStk->gtPutArgStkKind != GenTreePutArgStk::PutArgStkKindRepInstr));
+        if (preAdjustStack)
```

Comment:
I think it would be cleaner to extract the code from the beginning of `genPutArgStk` that determines whether it is using push (enhancing that logic as needed), and then in this method check `!m_pushStkArg` instead of using a local `preAdjustStack` variable. Then locally use `genStoreRegToStackArg` instead of having to duplicate the stack adjustment logic - then you'll only have the special casing of the "not a slot but treat as a slot" case.